rero_ils/modules/acquisition/rollover.py

```diff
@@ -0,0 +1,664 @@
+            o_order_pid = line.order_pid
+            p_order_pid = self.mapping_table.get('orders').get(o_order_pid)
+            if not p_order_pid:
+                raise RolloverError(
+                    f'Unable to find new parent order for {str(line)}'
+                    f' : parent pid was {p_order_pid}'
+                )
+            o_acc_pid = line.account_pid
+            p_acc_pid = self.mapping_table.get('accounts').get(o_acc_pid)
+            if not p_acc_pid:
+                raise RolloverError(
+                    f'Unable to find new parent account for {str(line)}'
+                    f' : parent pid was {p_acc_pid}'
+                )
+            data['acq_order']['$ref'] = get_ref_for_pid('acor', p_order_pid)
+            data['acq_account']['$ref'] = get_ref_for_pid('acac', p_acc_pid)
+            # Update specific order line fields
+            data['quantity'] = data.unreceived_quantity
+            del data['total_amount']
+
+            # Create the new order line.
+            #   If create failed :: raise an error.
+            #   If success :: fill the mapping table AND the stack of new obj.
+            try:
+                new_line = AcqOrderLine.create(
+                    data, dbcommit=True, reindex=True, delete_pid=True)
+                self.stack.append(new_line)
+                self.mapping_table['order_lines'][line.pid] = new_line.pid
+                old_label = truncate(str(line), 55).ljust(57)
+                new_label = truncate(str(new_line), 55)
+                log.info(f"\t* migrate {old_label} --> {new_label}")
+            except Exception as e:
+                raise RolloverError(f'Order line creation failed :: {str(e)}')\
+                    from e
+
+    def _update_budgets(self, orig_state=False, dest_state=False):
+        """Update rollover budgets to activate/deactivate them.
+
+        :param orig_state (boolean): the new state for original budget.
+        :param dest_state (boolean): the new state for destination budget.
+        """
+        self.logger.info("\tUpdating budget resources...")
+        orig_data = deepcopy(self.original_budget)
+        orig_data['is_active'] = orig_state
+        self.original_budget.update(orig_data, dbcommit=True, reindex=True)
+        state_str = 'activated' if orig_state else 'deactivated'
+        self.logger.info(f"\t  * Original budget is now {state_str}")
+
+        dest_data = deepcopy(self.destination_budget)
+        dest_data['is_active'] = dest_state
+        self.destination_budget.update(dest_data, dbcommit=True, reindex=True)
+        state_str = 'activated' if dest_state else 'deactivated'
+        self.logger.info(f"\t  * Destination budget is now {state_str}")
+
+    # PRIVATE METHODS =========================================================
+    #  These methods are used during the rollover process. They shouldn't be
+    #  use outside this class
+
+    def _abort_rollover(self, message=None):
```

Comment:
Is it possible to do that with a rollback or similar approach.

Reply:
It is the best solution  and we thought about it : do all operation in one db session and commit/rollback at the end.
The problem is that for acquisition resources, we use a lot of ES to compute/store some values : account encumbrance, account balance, order quantity received, ... theses data are only available through ES. And to index a record in ES, this record must be in DB.

It's why we use a stack of created objects. If errors and/or abort, we "just" need to delete all objects from this stack. Problems with this method are : slower and loss of PID.
It seems than "slow" isn't so important for the rollover script and the loss of pid isn't a real problem.